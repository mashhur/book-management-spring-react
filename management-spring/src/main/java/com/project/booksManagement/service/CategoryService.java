package com.project.booksManagement.service;

import java.util.List;

import com.project.booksManagement.model.Category;

public interface CategoryService {
	public Category addCategory(Category category) ;
	public Category updteCategory(Category category,Integer id);
	public void deleteCategory(Integer id);
	public List<Category> viewCategories();
	public Category getCategory(Integer id);

}
