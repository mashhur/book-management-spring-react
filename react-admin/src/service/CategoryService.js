const axios = require('axios');

const BASE_URL = 'http://localhost:8080/api/category';

export const getAllCategories = () => axios.get(BASE_URL);
export const getCategory = () => axios.get(BASE_URL);
export const updateCategory = (id, data) => axios.put(`${BASE_URL}/update/${id}`, data);
export const deleteCategory = (id) => axios.delete(`${BASE_URL}/${id}`);
export const createCategory = (name) => axios.post(`${BASE_URL}/add`, name);
