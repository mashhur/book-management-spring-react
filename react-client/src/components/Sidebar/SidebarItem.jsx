import { BiChevronRight } from "react-icons/bi";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";

function SidebarItem({ name, url }) {
	return (
		<div
			data-test="component-side-bar-item"
			style={{
				borderRadius: "2.3rem",
				fontSize: "1.0rem",
				marginBottom: ".5rem",
			}}
			className="col">
			<Link
				to={url ? url : `/cat/${name}`}
				className="list-group-item d-flex 
				justify-content-between 
				align-items-center 
				list-group-item-action 
				py-3 px-4 bg-category fs-6 ">
				{name}
				<span>
					<BiChevronRight
						style={{ fontSize: "23px", opacity: 0.6 }}
					/>
				</span>
			</Link>
		</div>
	);
}

SidebarItem.propType = {
	name: PropTypes.string.isRequired,
	url: PropTypes.string,
};

export default SidebarItem;
