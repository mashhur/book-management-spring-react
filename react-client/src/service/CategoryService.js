const axios = require("axios");

const BASE_URL = "http://localhost:8080/api/category";

class CategoryService {
	async getAllCategories() {
		return axios.get(BASE_URL);
	}
}

module.exports = new CategoryService();
