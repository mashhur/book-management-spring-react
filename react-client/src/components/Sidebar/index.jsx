import React, { useState } from "react";
import CategoryService from "../../service/CategoryService";
import SidebarItem from "./SidebarItem";

function Sidebar() {
	const [sidebarSate, setSideBarState] = React.useState({
		categories: [],
		isLoading: true,
		error: null,
	});

	const { categories, isLoading, error } = sidebarSate;

	const setCategories = (categories) => {
		setSideBarState({
			...sidebarSate,
			categories,
			isLoading: false,
		});
	};

	React.useEffect(() => {
		CategoryService.getAllCategories()
			.then((res) => {
				setCategories(res.data);
			})
			.catch((err) => {
				console.log("Error occured in Sidebar");
				setSideBarState({
					...sidebarSate,
					error: err.message,
					isLoading: false,
				});
			});
	}, []);

	return (
		<aside
			data-test="component-side-bar"
			style={{
				overflowY: "scroll",
				paddingBottom: "2rem",
				height: "89vh",
				background: "#f9f9f9a6",
			}}
			className="col-md-3 border px-0">
			<h4 className="h5 text-muted py-4 px-3 mb-3 border-bottom">
				Categories
			</h4>
			<p className="w-75 px-4 fs-6 text-muted">
				A curated list of every book ever written
			</p>

			<div className=" px-2  ">
				{isLoading ? (
					// isLoading
					<div
						data-test="loading-spinner"
						className="text-center">
						<div
							className="spinner-border text-primary"
							role="status"></div>
					</div>
				) : error ? (
					// showError
					<div
						data-test="error-message"
						className="text-center">
						<div
							className="alert alert-danger"
							role="alert">
							{error}
						</div>
					</div>
				) : (
					<div>
						<SidebarItem name="All Books" url="/" />
						{categories.map((category) => (
							<SidebarItem
								name={category.name}
								key={category.id}
							/>
						))}
					</div>
				)}
			</div>
		</aside>
	);
}

export default Sidebar;
