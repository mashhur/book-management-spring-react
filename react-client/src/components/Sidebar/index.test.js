import React from "react";
import { shallow } from "enzyme";
import { findByTestAttr } from "../../tests/testUtils";
import moxios from "moxios";
import CategoryService from "../../service/CategoryService";
import SidebarItem from "./SidebarItem";
import Sidebar from ".";

const setup = () => {
	return shallow(<Sidebar />);
};

describe("Testing side nav component", () => {
	beforeAll(() => {
		moxios.install();
	});
	afterAll(() => {
		moxios.uninstall();
	});

	it("renders without any error", async () => {
		const wrapper = setup();
		const component = await findByTestAttr(
			wrapper,
			"component-side-bar"
		);
		expect(component.length).toBe(1);
	});

	describe("when `getAllCategories()` is called", () => {
		it("return the correct repsonse", async () => {
			await moxios.wait(() => {
				const req = moxios.requests.mostRecent();
				req.respondWith({
					status: 200,
					response: [{ name: "Romance", id: 1 }],
				});
			});
			return CategoryService.getAllCategories().then(
				(respsonse) => {
					expect(respsonse.data).toEqual([
						{ name: "Romance", id: 1 },
					]);
				}
			);
		});

		//response error has been tested.
		it("return error response", async () => {
			await moxios.wait(() => {
				const req = moxios.requests.mostRecent();
				const errResp = {
					status: 500,
					response: {
						msg: "The server is unable to respond",
					},
				};
				req.reject(errResp);
			});

			return CategoryService.getAllCategories().catch(
				(error) => {
					// eslint-disable-next-line jest/no-conditional-expect
					expect(error.status).toEqual(500);
				}
			);
		});
	});

	describe("when an api call is made", () => {
		it("should call fn once only", async () => {
			jest
				.spyOn(React, "useEffect")
				.mockImplementationOnce((f) => f());
			jest.spyOn(CategoryService, "getAllCategories");
			setup();
			expect(
				CategoryService.getAllCategories
			).toHaveBeenCalledTimes(1);
		});

		// test loading state
		describe("while loading", () => {
			let wrapper;
			beforeEach(() => {
				const initalValue = {
					categories: [],
					isLoading: true,
					error: null,
				};

				React.useState = jest.fn(() => [
					initalValue,
					jest.fn(),
				]);

				wrapper = setup();
			});
			it("should render loading", async () => {
				const component = await findByTestAttr(
					wrapper,
					"loading-spinner"
				);
				expect(component.length).toBe(1);
			});

			it("should match the default loading snapshot", () => {
				expect(wrapper).toMatchSnapshot();
			});
		});

		//testing error state
		describe("when request fails", () => {
			let wrapper;
			beforeEach(() => {
				const initalValue = {
					categories: [],
					isLoading: false,
					error: "An error has occured",
				};

				React.useState = jest.fn(() => [
					initalValue,
					jest.fn(),
				]);

				wrapper = setup();
			});
			it("should render error msg", async () => {
				const wrapper = setup();
				const component = await findByTestAttr(
					wrapper,
					"error-message"
				);
				expect(component.length).toBe(1);
			});

			it("should match the error snapshot", () => {
				expect(wrapper).toMatchSnapshot();
			});
		});

		//testing categories state
		describe("when request is success", () => {
			let wrapper;
			beforeEach(() => {
				const initalValue = {
					categories: [
						{
							name: "Romance",
							id: 1,
						},
					],
					isLoading: false,
					error: null,
				};

				React.useState = jest.fn(() => [
					initalValue,
					jest.fn(),
				]);

				wrapper = setup();
			});
			it("should render items", async () => {
				expect(wrapper.find(SidebarItem).length).toBe(2);
			});

			it("should match the success snapshot", () => {
				expect(wrapper).toMatchSnapshot();
			});
		});
	});
});

//changes for this test
//TODO:change the filenames --done
//TODO: beforeAll and afterAll for moxios --done
//TODO: api call error catch --done
//TODO:Shift all data to another file for readibility --done
//TODO:snapshot testing for ui --done
//TODO:complete the incomplete tests --done
