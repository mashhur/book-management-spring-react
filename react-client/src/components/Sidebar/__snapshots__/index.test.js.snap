// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing side nav component when an api call is made when request fails should match the error snapshot 1`] = `ShallowWrapper {}`;

exports[`Testing side nav component when an api call is made when request is success should match the success snapshot 1`] = `ShallowWrapper {}`;

exports[`Testing side nav component when an api call is made while loading should match the default loading snapshot 1`] = `ShallowWrapper {}`;
